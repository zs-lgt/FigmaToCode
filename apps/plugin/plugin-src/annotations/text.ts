import { BaseAnnotation } from './base';

export class TextAnnotation extends BaseAnnotation {
  public async create(node: SceneNode, text: string = "请输入交互描述..."): Promise<void> {
    try {
      // 增加计数器
      this.state.annotationCounter++;

      // 创建源标注
      const sourceNumber = await this.createSourceAnnotation(node, this.state.annotationCounter);

      // 创建一个 frame 作为输入框容器
      const frame = figma.createFrame();
      frame.name = "标注框";
      frame.fills = [{ type: "SOLID", color: { r: 0.1, g: 0.1, b: 0.1 } }];
      frame.strokes = [{ type: "SOLID", color: { r: 0.2, g: 0.2, b: 0.2 } }];
      frame.strokeWeight = 1;
      frame.cornerRadius = 8;
      frame.clipsContent = false;
      frame.layoutMode = "VERTICAL";
      frame.primaryAxisSizingMode = "AUTO";
      frame.counterAxisSizingMode = "FIXED";
      frame.counterAxisAlignItems = "MIN";
      frame.primaryAxisAlignItems = "MIN";
      frame.paddingLeft = 16;
      frame.paddingRight = 16;
      frame.paddingTop = 16;
      frame.paddingBottom = 16;

      // 创建文本节点
      const textNode = figma.createText();
      textNode.name = "标注文本";
      textNode.fontSize = 14;
      textNode.fills = [{ type: "SOLID", color: { r: 1, g: 1, b: 1 } }];
      textNode.characters = text;

      // 设置文本节点大小和行为
      textNode.textAutoResize = "HEIGHT";
      textNode.resize(400 - 32, 400 - 32);
      textNode.textAlignVertical = "TOP";
      textNode.textAlignHorizontal = "LEFT";

      // 将文本添加到 frame 中
      frame.appendChild(textNode);

      // 创建标注框的圆形标记
      const targetNumber = figma.createFrame();
      targetNumber.name = "目标标记";
      targetNumber.resize(24, 24);
      targetNumber.cornerRadius = 12;
      targetNumber.fills = [{ type: "SOLID", color: { r: 1, g: 0, b: 0 } }];
      targetNumber.effects = [
        {
          type: "DROP_SHADOW",
          color: { r: 1, g: 1, b: 1, a: 1 },
          offset: { x: 0, y: 0 },
          radius: 4,
          spread: 2,
          visible: true,
          blendMode: "NORMAL",
        },
      ];

      // 创建目标标记的数字
      const targetLabel = figma.createText();
      targetLabel.name = "目标标记数字";
      targetLabel.fontSize = 14;
      targetLabel.fills = [{ type: "SOLID", color: { r: 1, g: 1, b: 1 } }];
      targetLabel.characters = this.state.annotationCounter.toString();
      targetLabel.textAlignHorizontal = "CENTER";
      targetLabel.textAlignVertical = "CENTER";
      targetLabel.x = (24 - targetLabel.width) / 2;
      targetLabel.y = (24 - targetLabel.height) / 2;

      targetNumber.appendChild(targetLabel);

      // 添加阴影效果
      frame.effects = [
        {
          type: "DROP_SHADOW",
          color: { r: 0, g: 0, b: 0, a: 0.25 },
          offset: { x: 0, y: 4 },
          radius: 8,
          spread: 0,
          visible: true,
          blendMode: "NORMAL",
        },
      ];

      // 设置插件数据
      frame.setPluginData("type", "annotation-frame");
      textNode.setPluginData("type", "annotation-text");
      sourceNumber.setPluginData("type", "annotation-number");
      targetNumber.setPluginData("type", "annotation-number");
      frame.setPluginData("annotation-number", this.state.annotationCounter.toString());

      // 设置初始尺寸和位置
      frame.resize(400, 400);
      frame.x = 30;
      frame.y = 0;

      // 设置标注框标记的位置
      targetNumber.x = 0;
      targetNumber.y = 0;

      // 创建一个组来包含标注框和其标记
      const annotationGroup = figma.group(
        [frame, targetNumber],
        figma.currentPage
      );
      annotationGroup.name = `标注组-${this.state.annotationCounter}/${node.name}/${node.id}`;

      // 获取节点的位置和大小
      const bounds = node.absoluteBoundingBox;
      if (!bounds) {
        throw new Error("无法获取节点边界");
      }

      // 设置组的位置和大小
      annotationGroup.resize(430, 400);
      annotationGroup.x = bounds.x + bounds.width + 50;
      annotationGroup.y = bounds.y;

      // 创建一个组来包含源标注
      const sourceGroup = figma.group([sourceNumber], figma.currentPage);
      sourceGroup.name = `源标注组-${this.state.annotationCounter}/${node.name}/${node.id}`;

      // 保存原始节点的引用
      sourceGroup.setPluginData("source-node-id", node.id);
      this.state.setSourceMarker(sourceGroup.id, node.id);

      // 建立双向关联
      sourceGroup.setPluginData("annotation-group-id", annotationGroup.id);
      annotationGroup.setPluginData("source-group-id", sourceGroup.id);

      // 刷新缓存
      this.state.invalidateCache();

      // 监听文本变化以调整高度
      const documentChangeHandler = this.eventManager.debounce((event: DocumentChangeEvent) => {
        if (this.eventManager.isLocalEvent(event)) return;
        const changes = event.documentChanges;

        // 处理文本变化
        const hasTextChange = changes.some(
          (change) =>
            change.type === "PROPERTY_CHANGE" &&
            change.node.id === textNode.id &&
            change.properties.includes("characters")
        );

        if (hasTextChange) {
          // 强制重新计算布局
          frame.layoutMode = "HORIZONTAL";
          frame.layoutMode = "VERTICAL";

          // 确保最小高度
          const contentHeight = textNode.height + 32; // 文本高度 + padding
          const newHeight = Math.max(400, contentHeight);

          // 如果高度需要调整，则调整
          if (frame.height !== newHeight) {
            frame.resize(400, newHeight);
            annotationGroup.resize(430, newHeight);
          }
        }
      }, 100);

      figma.on("documentchange", documentChangeHandler);

      // 监听选择变化
      const selectionChangeHandler = () => {
        const selection = figma.currentPage.selection;
        if (selection.length === 0 || !selection.includes(frame)) {
          if (textNode.characters === "") {
            textNode.characters = "请输入交互描述...";
          }
          // 只移除文本相关的事件监听器
          figma.off("documentchange", documentChangeHandler);
          figma.off("selectionchange", selectionChangeHandler);
        }
      };

      figma.on("selectionchange", selectionChangeHandler);

      // 保存计数器状态
      figma.clientStorage.setAsync("annotationCounter", this.state.annotationCounter);
      figma.notify("标注创建成功");
    } catch (error) {
      console.error("创建文本标注时出错:", error);
    }
  }
} 