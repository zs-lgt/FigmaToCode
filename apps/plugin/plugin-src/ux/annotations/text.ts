import { BaseAnnotation } from './base';

const TEXT_PADDING_Y = 12;
const TEXT_PADDING_X = 16;
const TEXT_DEFAULT_SIZE = 14;
const MIN_FRAME_HEIGHT = TEXT_DEFAULT_SIZE + TEXT_PADDING_Y * 2;
const isBase64 = (str: string): boolean => {
  return /^data:image\/(png|jpeg|jpg|gif|svg|webp|svg\+xml);base64,/.test(str);
}

export class TextAnnotation extends BaseAnnotation {
  public async createMultiContent(frame: FrameNode, content: string[]) {
    for(const item of content) {
      console.log('item', item);
      if (isBase64(item)) {
        // 其他图片格式直接处理
        await this.createImageNode(frame, item);
      } else {
        await this.createTextNode(frame, item);
      }
    }
  }

  public async createImageNode(frame: FrameNode, base64Data: string) {
    try {
      // Remove the data URL prefix to get the base64 string
      const base64String = base64Data.split(',')[1];
      // Convert base64 to array buffer
      const imageData = figma.base64Decode(base64String);
      // Create an image from the array buffer
      const image = await figma.createImage(imageData);
      // 获取图片尺寸
      const imageSize = await image.getSizeAsync();
      console.log('imageSize', imageSize)
      
      // Create rectangle that uses the image as a fill
      const rect = figma.createRectangle();
      rect.name = "标注图片";
      rect.fills = [{
        type: 'IMAGE',
        scaleMode: 'FILL',
        imageHash: image.hash
      }];
      
      // 设置图片尺寸
      if(imageSize.width > 0 && imageSize.height > 0) {
        rect.resize(imageSize.width, imageSize.height);
      }
      // Add the rectangle to the frame
      frame.appendChild(rect);
    } catch (error) {
      console.error("创建图片节点时出错:", error);
    }
  }

  public async createTextNode(frame: FrameNode, text: string) {
    const textNode = figma.createText();
    textNode.name = "标注文本";
    textNode.fontSize = 14;
    textNode.fills = [{ type: "SOLID", color: { r: 0.2, g: 0.2, b: 0.2 } }];
    textNode.characters = text;

    // 设置文本节点大小和行为
    textNode.textAutoResize = "HEIGHT";
    textNode.resize(400 - 32, textNode.fontSize + 8);
    textNode.textAlignVertical = "TOP";
    textNode.textAlignHorizontal = "LEFT";
    // 设置文本高度为自适应高度
    textNode.textAutoResize = "HEIGHT";

    // 将文本添加到 frame 中
    frame.appendChild(textNode);
  }


  // 找到节点的根层级
  private findRootParent(node: SceneNode): SceneNode {
    let current = node;
    while (current.parent && current.parent.type !== 'PAGE' && current.parent.type !== 'DOCUMENT') {
      current = current.parent;
    }
    return current;
  }

  /**
   * 调整标注位置以避免重叠
   * @param annotationContainer 标注容器节点
   * @param direction 防重叠方向，'vertical'表示纵向调整，'horizontal'表示横向调整
   * @param step 每次移动的步长（像素）
   */
  private adjustAnnotationPosition(annotationContainer: FrameNode, direction: 'vertical' | 'horizontal' = 'vertical', step: number = 20): void {
    const currentBounds = {
      x: annotationContainer.x,
      y: annotationContainer.y,
      width: annotationContainer.width,
      height: annotationContainer.height
    };
    let hasOverlap = true;
    
    while (hasOverlap) {
      hasOverlap = false;
      // 使用缓存获取所有现有的标注
      const existingAnnotations = this.state.getCategoryNodeCache().textAnnotations
        .filter(node => node.id !== annotationContainer.id);
      for (const existing of existingAnnotations) {
        const existingBounds = {
          x: existing.x,
          y: existing.y,
          width: existing.width,
          height: existing.height
        };
        
        // 检查是否有重叠
        if (this.checkOverlap(currentBounds, existingBounds)) {
          if (direction === 'vertical') {
            // 纵向调整：向下移动
            annotationContainer.y += step;
            currentBounds.y += step;
          } else {
            // 横向调整：向右移动
            annotationContainer.x += step;
            currentBounds.x += step;
          }
          hasOverlap = true;
          break;
        }
      }
    }
  }

  // 检查两个边界框是否重叠
  private checkOverlap(bounds1: { x: number, y: number, width: number, height: number },
                      bounds2: { x: number, y: number, width: number, height: number }): boolean {
    return !(bounds1.x + bounds1.width < bounds2.x ||
             bounds2.x + bounds2.width < bounds1.x ||
             bounds1.y + bounds1.height < bounds2.y ||
             bounds2.y + bounds2.height < bounds1.y);
  }

  public async create(node: SceneNode, text: string | string[] = "请输入交互描述..."): Promise<void> {
    try {
      // 增加计数器
      this.state.annotationCounter++;

      // 创建源标注
      const sourceNumber = await this.createSourceAnnotation(node, this.state.annotationCounter);

      // 创建一个 frame 作为输入框容器
      const frame = figma.createFrame();
      frame.name = "标注框";
      frame.fills = [{ type: "SOLID", color: { r: 0.98, g: 0.98, b: 0.98 } }];
      frame.strokes = [{ type: "SOLID", color: { r: 0.91, g: 0.91, b: 0.91 } }];
      frame.strokeWeight = 1;
      frame.cornerRadius = 4;
      frame.clipsContent = false;
      frame.layoutMode = "VERTICAL";
      frame.primaryAxisSizingMode = "AUTO";
      frame.counterAxisSizingMode = "AUTO";
      frame.counterAxisAlignItems = "MIN";
      frame.primaryAxisAlignItems = "MIN";
      frame.layoutSizingVertical = "HUG";
      frame.itemSpacing = 8;
      frame.paddingLeft = TEXT_PADDING_X;
      frame.paddingRight = TEXT_PADDING_X;
      frame.paddingTop = TEXT_PADDING_Y;
      frame.paddingBottom = TEXT_PADDING_Y;


      if (typeof text === 'object' && text.length) {
        await this.createMultiContent(frame, text)
      } else {
        // 创建文本节点
        const textNode = figma.createText();
        textNode.name = "标注文本";
        textNode.fontSize = 14;
        textNode.fills = [{ type: "SOLID", color: { r: 0.2, g: 0.2, b: 0.2 } }];
        textNode.characters = typeof text === 'string' ? text : text.join('\n');

        // 设置文本节点大小和行为
        textNode.textAutoResize = "HEIGHT";
        textNode.resize(400 - 32, textNode.fontSize + 8);
        textNode.textAlignVertical = "TOP";
        textNode.textAlignHorizontal = "LEFT";
        // 设置文本高度为自适应高度
        textNode.textAutoResize = "HEIGHT";

        // 将文本添加到 frame 中
        frame.appendChild(textNode);
      }

      // 创建标注框的圆形标记
      const targetNumber = figma.createFrame();
      targetNumber.name = "目标标记";
      targetNumber.resize(24, 24);
      targetNumber.cornerRadius = 12;
      targetNumber.fills = [{ type: "SOLID", color: { r: 1, g: 0, b: 0 } }];
      targetNumber.effects = [
        {
          type: "DROP_SHADOW",
          color: { r: 1, g: 1, b: 1, a: 1 },
          offset: { x: 0, y: 0 },
          radius: 4,
          spread: 2,
          visible: true,
          blendMode: "NORMAL",
        },
      ];

      // 创建目标标记的数字
      const targetLabel = figma.createText();
      targetLabel.name = "目标标记数字";
      targetLabel.fontSize = 14;
      targetLabel.fills = [{ type: "SOLID", color: { r: 1, g: 1, b: 1 } }];
      targetLabel.characters = this.state.annotationCounter.toString();
      targetLabel.textAlignHorizontal = "CENTER";
      targetLabel.textAlignVertical = "CENTER";
      targetLabel.x = (24 - targetLabel.width) / 2;
      targetLabel.y = (24 - targetLabel.height) / 2;

      targetNumber.appendChild(targetLabel);

      // 添加阴影效果
      // frame.effects = [
      //   {
      //     type: "DROP_SHADOW",
      //     color: { r: 0, g: 0, b: 0, a: 0.25 },
      //     offset: { x: 0, y: 4 },
      //     radius: 8,
      //     spread: 0,
      //     visible: true,
      //     blendMode: "NORMAL",
      //   },
      // ];

      // 设置插件数据
      frame.setPluginData("type", "annotation-frame");
      // textNode.setPluginData("type", "annotation-text");
      sourceNumber.setPluginData("type", "annotation-number");
      targetNumber.setPluginData("type", "annotation-number");
      frame.setPluginData("annotation-number", this.state.annotationCounter.toString());

      // 设置初始尺寸和位置
      // frame.resize(400, 400);
      frame.x = 30;
      frame.y = 0;

      // 设置标注框标记的位置
      targetNumber.x = 0;
      targetNumber.y = 0;

      // 创建一个frame来包含标注框和其标记
      const annotationContainer = figma.createFrame();
      annotationContainer.name = `标注组-${this.state.annotationCounter}/${node.name}/${node.id}`;
      annotationContainer.fills = [];
      annotationContainer.layoutMode = "HORIZONTAL";
      annotationContainer.primaryAxisSizingMode = "FIXED";
      annotationContainer.counterAxisSizingMode = "FIXED";
      annotationContainer.layoutSizingVertical = 'HUG';
      annotationContainer.layoutSizingHorizontal = 'HUG';
      annotationContainer.itemSpacing = 8;
      
      annotationContainer.clipsContent = false;
      
      // 将标注框和标记添加到frame中
      annotationContainer.appendChild(targetNumber);
      annotationContainer.appendChild(frame);

      // 获取节点的位置和大小
      const bounds = node.absoluteBoundingBox;
      if (!bounds) {
        throw new Error("无法获取节点边界");
      }

      // 节点定位
      // 找到目标节点的根层级
      const rootNode = this.findRootParent(node);
      const rootBounds = {
        x: rootNode.x,
        y: bounds.y,
        width: rootNode.width,
        height: rootNode.height
      };
      // 设置初始位置：根节点右侧20px
      annotationContainer.x = rootBounds.x + rootBounds.width + 20;
      annotationContainer.y = rootBounds.y;
      // 检查并避免与现有标注的重叠
      this.adjustAnnotationPosition(annotationContainer, 'horizontal');

      // 创建一个组来包含源标注
      const sourceGroup = figma.group([sourceNumber], figma.currentPage);
      sourceGroup.name = `源标注组-${this.state.annotationCounter}/${node.name}/${node.id}`;

      // 保存原始节点的引用
      sourceGroup.setPluginData("source-node-id", node.id);
      this.state.setSourceMarker(sourceGroup.id, node.id);

      // 建立双向关联
      sourceGroup.setPluginData("annotation-group-id", annotationContainer.id);
      annotationContainer.setPluginData("source-group-id", sourceGroup.id);

      // 注册到状态管理器的缓存中
      this.state.registerTextAnnotation(annotationContainer, sourceGroup);
      // 刷新缓存
      this.state.invalidateCache();

      // 监听文本变化以调整高度
      // const documentChangeHandler = this.eventManager.debounce((event: DocumentChangeEvent) => {
      //   if (this.eventManager.isLocalEvent(event)) return;
      //   const changes = event.documentChanges;

      //   // 处理文本变化
      //   const hasTextChange = changes.some(
      //     (change) =>
      //       change.type === "PROPERTY_CHANGE" &&
      //       change.node.id === textNode.id &&
      //       change.properties.includes("characters")
      //   );

      //   if (hasTextChange) {
      //     // 强制重新计算布局
      //     frame.layoutMode = "HORIZONTAL";
      //     frame.layoutMode = "VERTICAL";

      //     // 确保最小高度
      //     const contentHeight = textNode.height + 32; // 文本高度 + padding
      //     const newHeight = Math.max(400, contentHeight);

      //     // 如果高度需要调整，则调整
      //     if (frame.height !== newHeight) {
      //       frame.resize(400, newHeight);
      //       annotationGroup.resize(430, newHeight);
      //     }
      //   }
      // }, 100);

      // figma.on("documentchange", documentChangeHandler);

      // // 监听选择变化
      // const selectionChangeHandler = () => {
      //   const selection = figma.currentPage.selection;
      //   if (selection.length === 0 || !selection.includes(frame)) {
      //     if (textNode.characters === "") {
      //       textNode.characters = "请输入交互描述...";
      //     }
      //     // 只移除文本相关的事件监听器
      //     figma.off("documentchange", documentChangeHandler);
      //     figma.off("selectionchange", selectionChangeHandler);
      //   }
      // };

      // figma.on("selectionchange", selectionChangeHandler);

      // // 保存计数器状态
      figma.clientStorage.setAsync("annotationCounter", this.state.annotationCounter);
      // figma.notify("标注创建成功");
    } catch (error) {
      console.error("创建文本标注时出错:", error);
    }
  }
} 