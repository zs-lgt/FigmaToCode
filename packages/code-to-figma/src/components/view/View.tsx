import * as React from 'react';
import { Frame, Rectangle, StyleSheet } from '../..';
import { RectangleProps } from '../rectangle/Rectangle';
import { FrameNodeProps } from '../frame/Frame';

export type ViewProps = FrameNodeProps | RectangleProps;

const View: React.FC<ViewProps> = props => {
    const style = (props.style && StyleSheet.flatten(props.style)) || {};
    if (props.children) {
        return (
            <Frame
                {...props}
                style={[
                    { backgroundColor: 'transparent' },
                    (process.env.REACT_FIGMA_WEB_DEFAULTS_ENABLED &&
                        ((style as any).display === 'flex' || (style as any).display === 'inline-flex') && {
                            flexDirection: 'row'
                        }) ||
                        undefined,
                    (process.env.REACT_FIGMA_WEB_DEFAULTS_ENABLED && { alignItems: 'stretch' }) || undefined,
                    style
                ]}
            />
        );
    } else {
        return <Rectangle {...(props as RectangleProps)} />;
    }
};

export { View };
