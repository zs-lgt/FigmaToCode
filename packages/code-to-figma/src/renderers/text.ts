import { baseNodeMixin } from '../mixins/baseNodeMixin';
import { geometryMixin } from '../mixins/geometryMixin';
import { layoutMixin } from '../mixins/layoutMixin';
import { saveStyleMixin } from '../mixins/saveStyleMixin';
import { propsAssign } from '../helpers/propsAssign';
import { exportMixin } from '../mixins/exportMixin';
import { TextProps } from '../components/text/Text';
import { blendMixin } from '../mixins/blendMixin';
import { isValidSize } from '../helpers/isValidSize';
import { isEqualFontStyle } from '../helpers/isEqualFontStyle';
import { sceneNodeMixin } from '../mixins/sceneNodeMixin';
import { uiApi } from '../rpc';
import { safeGetPluginData } from '../helpers/safeGetPluginData';
import { constraintsMixin } from '../mixins/constraintsMixin';
import { DEFAULT_FONT } from '../helpers/constants';

const textNodePropsAssign = propsAssign<TextProps, TextProps>(
    [
        'characters',
        'textAlignHorizontal',
        'textAlignVertical',
        'paragraphIndent',
        'paragraphSpacing',
        'autoRename',
        'fontSize',
        'textCase',
        'textDecoration',
        'letterSpacing',
        'lineHeight',
        'textStyleId',
        'hyperlink'
    ],
    {
        characters: '',
        textAlignHorizontal: 'LEFT',
        textAlignVertical: 'TOP',
        paragraphIndent: 0,
        paragraphSpacing: 0,
        autoRename: false,
        fontSize: 12,
        textCase: 'ORIGINAL',
        textDecoration: 'NONE',
        letterSpacing: { value: 0, unit: 'PIXELS' },
        lineHeight: { unit: 'AUTO' },
        hyperlink: null
    }
);

const defaultFont = DEFAULT_FONT;

export const text = (node: TextNode) => (props: TextProps & { loadedFont?: FontName; hasDefinedWidth?: boolean }) => {
    const textNode = node || props.node || figma.createText();

    baseNodeMixin(textNode)(props);
    saveStyleMixin(textNode)(props);
    layoutMixin(textNode)(props);
    geometryMixin(textNode)(props);
    exportMixin(textNode)(props);
    blendMixin(textNode)(props);
    sceneNodeMixin(textNode)(props);
    constraintsMixin(textNode)(props);

    const { loadedFont, fontName = defaultFont } = props;
    if (
        loadedFont &&
        fontName &&
        loadedFont.family === fontName.family &&
        isEqualFontStyle(loadedFont.style, fontName.style)
    ) {
        if (props.fontName) {
            textNode.fontName = loadedFont;
        }
        if (
            props.hasDefinedWidth &&
            isValidSize(props.width) &&
            isValidSize(textNode.height) &&
            !props.textAutoResize
        ) {
            textNode.resize(props.width, textNode.height);
            textNode.textAutoResize = 'HEIGHT';
        } else {
            textNode.textAutoResize = props.textAutoResize || 'WIDTH_AND_HEIGHT';
        }

        const oldCharacters = textNode.characters;
        const oldFontSize = textNode.fontSize;
        textNodePropsAssign(textNode)(props);
        if (oldCharacters !== textNode.characters || oldFontSize !== textNode.fontSize) {
            const reactId = safeGetPluginData('reactId')(textNode);
            if (reactId) {
                uiApi.updateYogaNode(reactId);
            }
        }
    }

    return textNode;
};
