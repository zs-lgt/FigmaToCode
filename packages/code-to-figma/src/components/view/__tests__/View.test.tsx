import * as React from 'react';
import { View } from '../View';
import { Text } from '../../text/Text';
import { createFigma } from 'figma-api-stub';
const renderer = require('react-test-renderer');

describe('<View />', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true
        });
    });

    it('View without children', () => {
        const tree = renderer.create(<View style={{ backgroundColor: '#ff0000' }} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('View with children', () => {
        const tree = renderer
            .create(
                <View style={{ flex: 1 }}>
                    <Text>text</Text>
                </View>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('View with children and rectangle styles', () => {
        const tree = renderer
            .create(
                <View style={{ flex: 1, backgroundColor: '#ff0000' }}>
                    <Text>text</Text>
                </View>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('View with layoutAlign', () => {
        const tree = renderer
            .create(
                <View
                    style={{ width: 80, height: 40, backgroundColor: '#ffaa97' }}
                    layoutMode="HORIZONTAL"
                    layoutAlign="MAX"
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('alignItems: "stretch" on web defaults mode', () => {
        process.env.REACT_FIGMA_WEB_DEFAULTS_ENABLED = '1';
        const tree = renderer
            .create(
                <View>
                    <Text>text</Text>
                </View>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
        delete process.env.REACT_FIGMA_WEB_DEFAULTS_ENABLED;
    });

    it('flexDirection: "row" when display: flex on web defaults mode', () => {
        process.env.REACT_FIGMA_WEB_DEFAULTS_ENABLED = '1';
        const tree = renderer
            .create(
                <View style={{ display: 'flex' } as any}>
                    <Text>text</Text>
                </View>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
        delete process.env.REACT_FIGMA_WEB_DEFAULTS_ENABLED;
    });
});
