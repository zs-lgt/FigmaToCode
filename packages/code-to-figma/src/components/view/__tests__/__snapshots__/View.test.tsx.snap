// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<View /> View with children 1`] = `
<frame
  backgrounds={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 0,
        },
        "opacity": 0,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "transparent",
      "flex": 1,
    }
  }
>
  <text
    fills={
      Array [
        Object {
          "color": Object {
            "b": 0,
            "g": 0,
            "r": 0,
          },
          "opacity": 1,
          "type": "SOLID",
        },
      ]
    }
    fontName={
      Object {
        "family": "Inter",
        "style": "Regular",
      }
    }
    innerRef={
      Object {
        "current": undefined,
      }
    }
    loadedFont={null}
    style={Object {}}
  >
    text
  </text>
</frame>
`;

exports[`<View /> View with children and rectangle styles 1`] = `
<frame
  backgrounds={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 1,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#ff0000",
      "flex": 1,
    }
  }
>
  <text
    fills={
      Array [
        Object {
          "color": Object {
            "b": 0,
            "g": 0,
            "r": 0,
          },
          "opacity": 1,
          "type": "SOLID",
        },
      ]
    }
    fontName={
      Object {
        "family": "Inter",
        "style": "Regular",
      }
    }
    innerRef={
      Object {
        "current": undefined,
      }
    }
    loadedFont={null}
    style={Object {}}
  >
    text
  </text>
</frame>
`;

exports[`<View /> View with layoutAlign 1`] = `
<rectangle
  fills={
    Array [
      Object {
        "color": Object {
          "b": 0.592156862745098,
          "g": 0.6666666666666666,
          "r": 1,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  height={40}
  innerRef={
    Object {
      "current": undefined,
    }
  }
  layoutAlign="MAX"
  layoutMode="HORIZONTAL"
  style={
    Object {
      "alignSelf": "flex-end",
      "backgroundColor": "#ffaa97",
      "flexDirection": "row",
      "height": 40,
      "width": 80,
    }
  }
  width={80}
/>
`;

exports[`<View /> View without children 1`] = `
<rectangle
  fills={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 1,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#ff0000",
    }
  }
/>
`;

exports[`<View /> alignItems: "stretch" on web defaults mode 1`] = `
<frame
  backgrounds={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 0,
        },
        "opacity": 0,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "transparent",
    }
  }
>
  <text
    fills={
      Array [
        Object {
          "color": Object {
            "b": 0,
            "g": 0,
            "r": 0,
          },
          "opacity": 1,
          "type": "SOLID",
        },
      ]
    }
    fontName={
      Object {
        "family": "Inter",
        "style": "Regular",
      }
    }
    innerRef={
      Object {
        "current": undefined,
      }
    }
    loadedFont={null}
    style={Object {}}
  >
    text
  </text>
</frame>
`;

exports[`<View /> flexDirection: "row" when display: flex on web defaults mode 1`] = `
<frame
  backgrounds={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 0,
        },
        "opacity": 0,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "alignItems": "stretch",
      "backgroundColor": "transparent",
      "display": "flex",
      "flexDirection": "row",
    }
  }
>
  <text
    fills={
      Array [
        Object {
          "color": Object {
            "b": 0,
            "g": 0,
            "r": 0,
          },
          "opacity": 1,
          "type": "SOLID",
        },
      ]
    }
    fontName={
      Object {
        "family": "Inter",
        "style": "Regular",
      }
    }
    innerRef={
      Object {
        "current": undefined,
      }
    }
    loadedFont={null}
    style={Object {}}
  >
    text
  </text>
</frame>
`;
